solvers/fista_automatic_restart.py

```diff
@@ -28,63 +24,66 @@ def set_objective(self, X, y, lmbd, fit_intercept):
         self.X, self.y, self.lmbd = X, y, lmbd
         self.fit_intercept = fit_intercept
 
-    def run(self, n_iter):
+    def run(self, callback):
         L = self.compute_lipschitz_constant()
         n_features = self.X.shape[1]
         w = np.zeros(n_features)
         z = np.zeros(n_features)
-        C = 6.38
-        ite_per_restart = np.zeros(int(n_iter / (2 * C)) + 2)
-        F_tab = np.zeros(int(n_iter / (2 * C)) + 2)
-        i_glob = 0
-        i_int = 0
-        counter = 0  # Restart counter
-        LastF = 0
-        ite_per_restart[counter] = int(2*C)
-        F_tab[counter] = (.5*np.linalg.norm(self.X @ w - self.y, ord=2)**2
-                          + self.lmbd*abs(w).sum())
+        C = 6.38  # Parameter controlling the restart frequency, 6.38 is the
+        # optimal theoretical value (we refer the user to the preprint for
+        # more details).
+        inner_iteration = 0
+        n_restarts = 0  # Restart counter
+        last_F = 0
+        ite_per_restart = np.array([int(2*C)])
+        F_tab = np.array([self.cost_function(w)])
+        cb = callback(w)
         # First call of FISTA
-        while i_glob < n_iter and i_int < ite_per_restart[counter]:
+        while cb and inner_iteration < ite_per_restart[n_restarts]:
             w_old = w.copy()
             z -= self.X.T @ (self.X @ z - self.y) / L
             w = self.st(z, self.lmbd / L)
-            z = w + i_int/(i_int+3) * (w - w_old)
-            i_int += 1
-            i_glob += 1
-        counter += 1
-        ite_per_restart[counter] = int(2*C)
-        F_tab[counter] = (.5*np.linalg.norm(self.X @ w - self.y, ord=2)**2
-                          + self.lmbd*abs(w).sum())
-        while i_glob < n_iter:
-            i_int = 0
+            z = w + inner_iteration/(inner_iteration+3) * (w - w_old)
+            inner_iteration += 1
+            cb = callback(w)
+        n_restarts += 1
+        ite_per_restart = np.r_[ite_per_restart, int(2*C)]
+        F_tab = np.r_[F_tab, self.cost_function(w)]
+        while cb:
+            inner_iteration = 0
             # Restart of FISTA after k_tab[counter] iterations
-            while i_glob < n_iter and i_int < ite_per_restart[counter]:
+            while cb and inner_iteration < ite_per_restart[n_restarts]:
```

Comment:
```suggestion
            while should_run and inner_iteration < ite_per_restart[n_restarts]:
```

Reply:
Also, a maybe easier to read pattern for such loop would be to use something like:
```python
for i in range(ite_per_restart[n_restart]):
    ...
    should_run = callback(w)
    if not should_run:
        break
```
this way the only `while` loop is on `should_run`, which avoid the use of manually increasing counters, except for `n_restart` (which can also probably be dropped, see next comments).